src/test/socket/sendto_recvfrom/test_sendto_recvfrom.rs

```diff
@@ -0,0 +1,1379 @@
+        Self {
+            fd: -1,
+            buf: None,
+            len: 0,
+            flags: 0,
+            addr: None,
+            addr_len: None,
+        }
+    }
+}
+
+fn main() -> Result<(), String> {
+    // should we restrict the tests we run?
+    let filter_shadow_passing = std::env::args().any(|x| x == "--shadow-passing");
+    let filter_libc_passing = std::env::args().any(|x| x == "--libc-passing");
+    // should we summarize the results rather than exit on a failed test
+    let summarize = std::env::args().any(|x| x == "--summarize");
+
+    let mut tests = get_tests();
+    if filter_shadow_passing {
+        tests = tests
+            .into_iter()
+            .filter(|x| x.passing(TestEnv::Shadow))
+            .collect()
+    }
+    if filter_libc_passing {
+        tests = tests
+            .into_iter()
+            .filter(|x| x.passing(TestEnv::Libc))
+            .collect()
+    }
+
+    test_utils::run_tests(&tests, summarize)?;
+
+    println!("Success.");
+    Ok(())
+}
+
+fn get_tests() -> Vec<test_utils::ShadowTest<(), String>> {
+    let mut tests: Vec<test_utils::ShadowTest<_, _>> = vec![];
+
+    tests.extend(vec![
+        test_utils::ShadowTest::new(
+            "test_invalid_fd",
+            test_invalid_fd,
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        test_utils::ShadowTest::new(
+            "test_non_existent_fd",
+            test_non_existent_fd,
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        test_utils::ShadowTest::new(
+            "test_non_socket_fd",
+            test_non_socket_fd,
+            set![TestEnv::Libc],
+        ),
+        test_utils::ShadowTest::new(
+            "test_not_connected_tcp",
+            test_not_connected_tcp,
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        test_utils::ShadowTest::new(
+            "test_not_connected_udp",
+            test_not_connected_udp,
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        test_utils::ShadowTest::new(
+            "test_null_buf <tcp>",
+            || test_null_buf(libc::SOCK_STREAM),
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        test_utils::ShadowTest::new(
+            "test_null_buf <udp>",
+            || test_null_buf(libc::SOCK_DGRAM),
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        test_utils::ShadowTest::new(
+            "test_zero_len_buf <tcp>",
+            || test_zero_len_buf(libc::SOCK_STREAM),
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        test_utils::ShadowTest::new(
+            "test_zero_len_buf <udp>",
+            || test_zero_len_buf(libc::SOCK_DGRAM),
+            set![TestEnv::Libc, TestEnv::Shadow],
+        ),
+        // This test fails in shadow (debug mode) on centos 7, but only in the
+        // GitHub CI, so disabling for now.
+        // The cause of failure seems to be that centos 7 is about twice as slow
```

Comment:
We could just increase the timeout for this test, as we do for phold.

Reply:
I increased it to 60 seconds. It seems to take about 40 seconds in the CI on CentOS 7 (and 3 seconds on my laptop).